{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "JMa9tKjTZ_fK"
   },
   "source": [
    "# Deep learning model to identify target mRNA of microRNA sequences\n",
    "\n",
    "From a review of currently developed Deep learning models in genomics, I realized there is still a long way to go before the full stock of machine learning techniques will be applied to genomics at its maximum potential. In particular, I got fascinated by the deepTarget model that has recently been proposed as a way to identify microRNA targets with 96% accuracy - therefore I decided to focus my project in this field. The repository of the mentioned work is available on GitHib at the following link: https://github.com/ailab-seoultech/deepTarget.\n",
    "\n",
    "MicroRNAs (miRNAs), which are small non-coding RNA molecules that consist of about 22 nucleotides, are known to regulate more than 60% of protein coding genes of humans and other mammals at the RNA level. As miRNAs control the function of their target messenger RNAs (mRNAs) by regulating the expression of the targets, investigating miRNAs is important to understand various biological processes, including diseases. To predict targets of given miRNAs, numerous computational tools have been proposed. \n",
    "\n",
    "## Problem Statement\n",
    "Two types of computational problems about miRNAs thus naturally arise in bioinformatics: miRNA host identification (i.e., the problem of locating the genes that encode pre-miRNAs) and miRNA target prediction (i.e., the task of finding the mRNA targets of given miRNAs). This project focuses on the target prediction problem, aiming at deploying a model that is able to recognize whether or not a sequence of microRNA belongs to a given target mRNA. The output will be a binary classification of miRNA-mRNA pairs that match/ don’t match (i.e. mRNA is a target of the given miRNA).\n",
    "\n",
    "\n",
    "## Dataset\n",
    "I will utilize experimental negative data collected from a 2020 paper ‘Deep Learning-Based microRNA Target Prediction Using Experimental Negative Data’ that used the same dataset to train a deepTarget model.\n",
    "\n",
    "The dataset is available here: https://github.com/ailab-seoultech/deepTarget \n",
    "\n",
    "The dataset is already split into ‘train’ and ‘test’ with 65,425 and 10,930 data points respectively. The train set has a list of labelled pairs of miRNA and mRNA sequences structured as follows:\n",
    "- MIRNA_ID: identifier of the miRNA sequence (note this is not a unique identifier of the miRNA-mRNA pair because a miRNA can be listed multiple times, paired with multiple mRNA targets)\n",
    "- ENSEMLBE_GENE and GENE_SYMBOL:  identifiers of the gene encoding such miRNA and mRNA sequences. This enables for gene-level predictions of mRNA-miRNA pairs\n",
    "- MIRNA_SEQ: list of strings with the sequence of basis of the miRNA (e.g. ‘UGAGGU…’)\n",
    "- MRNA_SEQ: list of strings with the sequence of basis of the mRNA (e.g. ‘GCCTGC…’)\n",
    "- LABEL: 0 or 1 if the mRNA is a target of the miRNA (1) or it is not (0)\n",
    "\n",
    "The dataset also includes 2 'fatsa' files, listing all possible MIRNA and MRNA IDs and their relative nucleotide sequences.\n",
    "\n",
    "The dataset has been generated using public data obtained from several datasets. It has been balanced by the creators by randomly sampling positive and negative pairs3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "eUjS-KPpSTFp"
   },
   "source": [
    "## Setup\n",
    "Let's first install the right version of the softwares to use and import the needed libraries "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "e2nY4cRANuZ7",
    "outputId": "bbda5459-28fa-4c2f-f510-ed0a31b4ea77",
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: biopython in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (1.78)\n",
      "Requirement already satisfied: numpy in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (from biopython) (1.19.5)\n",
      "Requirement already satisfied: torch in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (1.5.1)\n",
      "Requirement already satisfied: future in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (from torch) (0.18.2)\n",
      "Requirement already satisfied: numpy in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (from torch) (1.19.5)\n",
      "Requirement already satisfied: regex in /home/ec2-user/anaconda3/envs/amazonei_pytorch_latest_p36/lib/python3.6/site-packages (2020.7.14)\n",
      "\u001b[31mERROR: Could not find a version that satisfies the requirement nomkl\u001b[0m\n",
      "\u001b[31mERROR: No matching distribution found for nomkl\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install biopython\n",
    "!pip install torch\n",
    "!pip install regex\n",
    "!pip install nomkl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "iw-H4SwJkFyU"
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "import torch.optim as optim\n",
    "import torch.nn.functional as F\n",
    "from torch.utils.data import DataLoader\n",
    "\n",
    "from tqdm.auto import tqdm\n",
    "bar_format = '{desc} |{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}{postfix}]'\n",
    "import sklearn\n",
    "from sklearn.utils.class_weight import compute_class_weight\n",
    "from datetime import datetime\n",
    "\n",
    "import argparse\n",
    "import regex\n",
    "import Bio\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Exploration\n",
    "In the same location of the Jupyter notebook, I provided a 'data' folder downloaded from the GitHub repository of the paper. It includes 'fasta' files, a training set and 10 test sets\n",
    "### Fatsa files\n",
    ".fasta files include IDs and Sequences of all MRNAs and MIRNAs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hsa-miR-4777-5p\n",
      "UUCUAGAUGAGAGAUAUAUAUA\n",
      "hsa-miR-3908\n",
      "GAGCAAUGUAGGUAGACUGUUU\n",
      "hsa-miR-96-3p\n",
      "AAUCAUGUGCAGUGCCAAUAUG\n",
      "hsa-miR-3144-5p\n",
      "AGGGGACCAAAGAGAUAUAUAG\n",
      "hsa-miR-6509-3p\n",
      "UUCCACUGCCACUACCUAAUUU\n"
     ]
    }
   ],
   "source": [
    "for i in range (5):\n",
    "    print(list(SeqIO.parse(\"data/mirna.fasta\", 'fasta'))[i].id)\n",
    "    print(list(SeqIO.parse(\"data/mirna.fasta\", 'fasta'))[i].seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Training set\n",
    "train data include pairs of MRNAs and MIRNAs labelled if they are or not a match\n",
    "Tha dataset is balanced, including half of labels 0 and half of label 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIRNA_ID</th>\n",
       "      <th>ENSEMBL_GENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>MIRNA_SEQ</th>\n",
       "      <th>MRNA_SEQ</th>\n",
       "      <th>LABEL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENSG00000114573</td>\n",
       "      <td>ATP6V1A</td>\n",
       "      <td>UGAGGUAGUAGGUUGUAUAGUU</td>\n",
       "      <td>GCCTGCTATTGAGGAAAGGTATTCTTCTATACAACTTGTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENSG00000104497</td>\n",
       "      <td>SNX16</td>\n",
       "      <td>UGAGGUAGUAGGUUGUAUAGUU</td>\n",
       "      <td>ATTTGTTTAGTTTCCACGTAATCTTTATCTCTACCTAGAT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENSG00000141682</td>\n",
       "      <td>PMAIP1</td>\n",
       "      <td>UGAGGUAGUAGGUUGUAUAGUU</td>\n",
       "      <td>GCACATTGTATATGATTCGGTTTATACATATTACCTTGTT</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>ENSG00000174010</td>\n",
       "      <td>KLHL15</td>\n",
       "      <td>UGAGGUAGUAGGUUGUAUAGUU</td>\n",
       "      <td>GAAGTTAGACACCTTTCTGCTAGACAACTTTGTGCCACTC</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>ENSG00000130402</td>\n",
       "      <td>ACTN4</td>\n",
       "      <td>UGAGGUAGUAGGUUGUGUGGUU</td>\n",
       "      <td>GGCCCTCATCTTCGACAACAAGCACACCAACTATACCATG</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        MIRNA_ID     ENSEMBL_GENE GENE_SYMBOL               MIRNA_SEQ  \\\n",
       "0  hsa-let-7a-5p  ENSG00000114573     ATP6V1A  UGAGGUAGUAGGUUGUAUAGUU   \n",
       "1  hsa-let-7a-5p  ENSG00000104497       SNX16  UGAGGUAGUAGGUUGUAUAGUU   \n",
       "2  hsa-let-7a-5p  ENSG00000141682      PMAIP1  UGAGGUAGUAGGUUGUAUAGUU   \n",
       "3  hsa-let-7a-5p  ENSG00000174010      KLHL15  UGAGGUAGUAGGUUGUAUAGUU   \n",
       "4  hsa-let-7b-5p  ENSG00000130402       ACTN4  UGAGGUAGUAGGUUGUGUGGUU   \n",
       "\n",
       "                                   MRNA_SEQ  LABEL  \n",
       "0  GCCTGCTATTGAGGAAAGGTATTCTTCTATACAACTTGTT      1  \n",
       "1  ATTTGTTTAGTTTCCACGTAATCTTTATCTCTACCTAGAT      1  \n",
       "2  GCACATTGTATATGATTCGGTTTATACATATTACCTTGTT      1  \n",
       "3  GAAGTTAGACACCTTTCTGCTAGACAACTTTGTGCCACTC      1  \n",
       "4  GGCCCTCATCTTCGACAACAAGCACACCAACTATACCATG      1  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(\"data/train/train_set.csv\", header=0, sep='\\t')\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[<AxesSubplot:title={'center':'LABEL'}>]], dtype=object)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.hist(column='LABEL')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "bkSXa_wdaXAw"
   },
   "source": [
    "## Pre-processing\n",
    "Below a set of functions that will help me read and encode the .fasta files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "tyArlMNfkMr9",
    "outputId": "7076e8b1-90a7-4dcd-f66f-7a8a64d5e93e"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mirna_ID : hsa-miR-4777-5p. , mirna_SEQUENCE : UUCUAGAUGAGAGAUAUAUAUA\n",
      "mirna_ID : hsa-miR-3908. , mirna_SEQUENCE : GAGCAAUGUAGGUAGACUGUUU\n",
      "mirna_ID : hsa-miR-96-3p. , mirna_SEQUENCE : AAUCAUGUGCAGUGCCAAUAUG\n",
      "mirna_ID : hsa-miR-3144-5p. , mirna_SEQUENCE : AGGGGACCAAAGAGAUAUAUAG\n",
      "mirna_ID : hsa-miR-6509-3p. , mirna_SEQUENCE : UUCCACUGCCACUACCUAAUUU\n",
      "\n",
      "\n",
      "mrna_ID : NM_003629. , mrna_SEQUENCE : AGAGGAAGUGGGAAGAGAGGUGGUUCUCUGGCAUUUUUUU\n",
      "mrna_ID : NM_001135041. , mrna_SEQUENCE : GCACUCCUUUCCCCUGCUGUCCCCUUCGACCCUCAGCCCU\n",
      "mrna_ID : NM_001256461. , mrna_SEQUENCE : AUGUUCUAUACAGUGGACAGCCCUCCAGAAUGGUACUUCA\n",
      "mrna_ID : NM_005371. , mrna_SEQUENCE : CUGCUUACUCUACCUUAGCUGGACCUCGUCUCCCAGGGAU\n",
      "mrna_ID : NM_175734. , mrna_SEQUENCE : CCAGCAGGCGGAUGUGGGGUGUGGGGCAGGGCAUGGAGGG\n"
     ]
    }
   ],
   "source": [
    "def read_fasta(mirna_fasta_file, mrna_fasta_file):\n",
    "    # function to read the '.fatsa' files and retun sequences and ids of \n",
    "    # mRNA and miRNA\n",
    "    mirna_list = list(SeqIO.parse(mirna_fasta_file, 'fasta'))\n",
    "    mrna_list = list(SeqIO.parse(mrna_fasta_file, 'fasta'))\n",
    "    \n",
    "    mirna_ids = []\n",
    "    mirna_seqs = []\n",
    "    mrna_ids = []\n",
    "    mrna_seqs = []\n",
    "    \n",
    "    for i in range(len(mirna_list)):\n",
    "        mirna_ids.append(str(mirna_list[i].id))\n",
    "        mirna_seqs.append(str(mirna_list[i].seq))\n",
    "    \n",
    "    for i in range(len(mrna_list)):\n",
    "        mrna_ids.append(str(mrna_list[i].id))\n",
    "        mrna_seqs.append(str(mrna_list[i].seq))\n",
    "    \n",
    "    return mirna_ids, mirna_seqs, mrna_ids, mrna_seqs\n",
    "\n",
    "\n",
    "mirna_fasta_file = \"data/mirna.fasta\"\n",
    "mrna_fasta_file  = \"data/mrna.fasta\"\n",
    "mirna_ids, mirna_seqs, mrna_ids, mrna_seqs = read_fasta(mirna_fasta_file,mrna_fasta_file)\n",
    "\n",
    "for i in range(5):\n",
    "    print(f\"mirna_ID : {mirna_ids[i]}. , mirna_SEQUENCE : {mirna_seqs[i]}\")\n",
    "\n",
    "print(\"\\n\")\n",
    "\n",
    "for i in range(5):\n",
    "    print(f\"mrna_ID : {mrna_ids[i]}. , mrna_SEQUENCE : {mrna_seqs[i][:40]}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "g5ms9t5toV4h",
    "outputId": "f6a929ae-8081-4a4d-cf65-4bd3523f4d87"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3 2 1 2 4 2 2 4 4 4 2 2 2 2 4 3 2 4 3 4 2 2 2 2 4 4 2 3 1 2 2 2 4 2 1 3 2\n",
      " 2 2 4 2 4 3 3 4 3 2 2 3 2 4 2 4 3 2 2 2 3 1 4 3 2 1 2 1 3 2 2 1 2 2 4 2 1\n",
      " 3 2 2 1 3 2 2 2 2 2 1 3 3 4 1 3 1 1 1 2 3 4 3 3 3 4 4 1 1 3 2 4 2 4 4 2 2\n",
      " 4 3 2 2 2 2 3 4 4 2 1 3 2 4 4 2 1 2 4 2 2 2 1 2 2 2 4 4 4 2 1 3 2 3 4 2 2\n",
      " 4 3 2 2 2 2 4 4 2 1 2 2 4 4 3 1 2 2 2 3 3 3 4 4 2 2 2 2 2 1 2 4 2 2 2 1 4\n",
      " 4 2 2 2 4 3 3 2 2 4 2 4 3 2 2 1 4 1 1 4 4 4 3 4 4 3 4 4 2 1 1 2 4 3 2 4 2\n",
      " 2 2 4 2 2 4 4 2 2 4 3 1 3 3 3 3 2 2 4 2 1 3 3 3 2 4 4 3 4 3 3 3 3 3 3 4 1\n",
      " 3 3 2 4 3 1 3 1 2 2 2 2 1 2 2 1 2 2 1 1 1 3 3 4 4 1 1 3 4 3 1 3 3 4 2 2 2\n",
      " 2 4 4 3 1 4 4 3 1 3 3 1 2 4 4 2 1 2 2 2 2 4 4 3 1 4 4 1 1 1 3 2 1 1 2 4 4\n",
      " 2 4 3 2 4 4 2 1 3 4 3 2]\n"
     ]
    }
   ],
   "source": [
    "def nucleotide_to_int(nucleotides, max_len):\n",
    "    # assign a number to the basis in order to translate nucleotides into int\n",
    "    dictionary = {'A':1, 'C':2, 'G':3, 'T':4, 'U':4}\n",
    "    \n",
    "    chars = []\n",
    "    nucleotides = nucleotides.upper()\n",
    "    for c in nucleotides:\n",
    "        chars.append(c)\n",
    "    \n",
    "    ints_enc = np.full((max_len,), fill_value=0) # to post-pad inputs\n",
    "    for i in range(len(chars)):\n",
    "        try:\n",
    "            ints_enc[i] = dictionary[chars[i]]\n",
    "        except KeyError:\n",
    "            continue\n",
    "        except IndexError:\n",
    "            break\n",
    "        \n",
    "    return ints_enc\n",
    "\n",
    "\n",
    "seq = mrna_seqs[1]\n",
    "max_len = len(seq)\n",
    "\n",
    "ints_enc = nucleotide_to_int(seq,max_len)\n",
    "\n",
    "print(ints_enc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "W0SxQqahy8U_",
    "outputId": "ffdaf2a7-8c10-4e47-a1a2-d3f08ed5b72a"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[3 2 1 2 4 2 2 4 4 4]\n"
     ]
    }
   ],
   "source": [
    "def sequence_to_int(sequences, max_len):\n",
    "    # translate entire RNA sequences into int\n",
    "    import itertools\n",
    "    \n",
    "    if type(sequences) is list:\n",
    "        seqs_enc = np.asarray([nucleotide_to_int(seq, max_len) for seq in sequences])\n",
    "    else:\n",
    "        seqs_enc = np.asarray([nucleotide_to_int(seq, max_len) for seq in sequences])\n",
    "        seqs_enc = list(itertools.chain(*seqs_enc))\n",
    "        seqs_enc = np.asarray(seqs_enc)\n",
    "    \n",
    "    return seqs_enc\n",
    "\n",
    "x=sequence_to_int(mrna_seqs, 10)\n",
    "print (x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "c1vLw5DDzrat",
    "outputId": "29da04e4-d211-431f-9e8d-c31838a78009"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0\n",
      " 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 3 2 1 2 4 2 2 4 4 4]\n"
     ]
    }
   ],
   "source": [
    "def pad_sequences(sequences, max_len=None, padding='pre', fill_value=0):\n",
    "    # padding sequences with fill_value to achieve seme lenght\n",
    "    n_samples = len(sequences)\n",
    "    \n",
    "    lengths = []\n",
    "    for seq in sequences:\n",
    "        try:\n",
    "            lengths.append(len(seq))\n",
    "        except TypeError:\n",
    "            raise ValueError(\"sequences expected a list of iterables, got {}\".format(seq))\n",
    "    if max_len is None:\n",
    "        max_len = np.max(lengths)\n",
    "    \n",
    "    input_shape = np.asarray(sequences[0]).shape[1:]\n",
    "    padded_shape = (n_samples, max_len) + input_shape\n",
    "    padded = np.full(padded_shape, fill_value=fill_value)\n",
    "    \n",
    "    for i, seq in enumerate(sequences):\n",
    "        if padding == 'pre':\n",
    "            truncated = seq[-max_len:]\n",
    "            padded[i, -len(truncated):] = truncated\n",
    "        elif padding == 'post':\n",
    "            truncated = seq[:max_len]\n",
    "            padded[i, :len(truncated)] = truncated\n",
    "        else:\n",
    "            raise ValueError(\"padding expected 'pre' or 'post', got {}\".format(truncating))\n",
    "    \n",
    "    return padded\n",
    "\n",
    "x=pad_sequences(x, max_len=100)\n",
    "print (x[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "K68XiRph0hLd",
    "outputId": "d1f8f6b9-278f-474a-bad7-161d0c2b7c1f"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[0.0, 1.0, 0.0, 0.0], [0.0, 1.0, 0.0, 0.0], [0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0], [0.0, 0.0, 0.0, 1.0]]\n"
     ]
    }
   ],
   "source": [
    "def one_hot(ints):\n",
    "    #one hot encoding for nucleotides\n",
    "    dictionary_k = 5 # maximum number of nucleotides\n",
    "    ints_len = len(ints)\n",
    "    ints_enc = np.zeros((ints_len, dictionary_k))\n",
    "    ints_enc[np.arange(ints_len), [k for k in ints]] = 1\n",
    "    ints_enc = ints_enc[:, 1:5] # to handle zero-padded values\n",
    "    ints_enc = ints_enc.tolist()\n",
    "    \n",
    "    return (ints_enc)\n",
    "\n",
    "def one_hot_enc(seqs_enc):\n",
    "    #one hot encoding for sequences    \n",
    "    one_hot_encs = []\n",
    "    \n",
    "    for i in range(len(seqs_enc)):\n",
    "        one_hot_encs.append(one_hot(seqs_enc[i]))\n",
    "    \n",
    "    one_hot_encs = np.array(one_hot_encs)\n",
    "    \n",
    "    return one_hot_encs\n",
    "\n",
    "b=one_hot(x[1])\n",
    "print (b[95:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "CRkECXQ733ld",
    "outputId": "9657aa23-5f3a-4fb2-9ce8-b64d8daf3bd1"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[1. 0. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 0.]\n",
      " [1. 0. 0. 0. 0.]\n",
      " [0. 0. 1. 0. 0.]\n",
      " [0. 0. 0. 1. 0.]\n",
      " [0. 0. 0. 0. 1.]\n",
      " [1. 0. 0. 0. 0.]\n",
      " [0. 1. 0. 0. 0.]\n",
      " [0. 0. 1. 0. 0.]\n",
      " [0. 0. 0. 1. 0.]]\n"
     ]
    }
   ],
   "source": [
    "def to_categorical(labels, n_classes=None):\n",
    "    #matrix assigning labels to a number of classes\n",
    "    labels = np.array(labels, dtype='int').reshape(-1)\n",
    "\n",
    "    n_samples = labels.shape[0]\n",
    "    if not n_classes:\n",
    "        n_classes = np.max(labels) + 1\n",
    "\n",
    "    categorical = np.zeros((n_samples, n_classes))\n",
    "    categorical[np.arange(n_samples), labels] = 1\n",
    "    \n",
    "    return categorical\n",
    "\n",
    "labels=[0,1,1,1,1,0,2,3,4,0,1,2,3]\n",
    "labels=to_categorical(labels)\n",
    "print(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "id": "hdNFVjVinHjf"
   },
   "outputs": [],
   "source": [
    "def preprocess_data(x_query_seqs, x_target_seqs, y=None, cts_size=None, pre_padding=False):\n",
    "    #getting encoded data form miRNA and mRNA sequences\n",
    "    if cts_size is not None:\n",
    "        max_len = cts_size\n",
    "    else:\n",
    "        max_len = max(len(max(x_query_seqs, key=len)), len(max(x_target_seqs, key=len)))\n",
    "    \n",
    "    x_mirna = sequence_to_int(x_query_seqs, max_len)\n",
    "    x_mrna = sequence_to_int(x_target_seqs, max_len)\n",
    "    \n",
    "    if pre_padding:\n",
    "        x_mirna = pad_sequences(x_mirna, max_len, padding='pre')\n",
    "        x_mrna = pad_sequences(x_mrna, max_len, padding='pre')\n",
    "    \n",
    "    x_mirna_embd = one_hot_enc(x_mirna)\n",
    "    x_mrna_embd = one_hot_enc(x_mrna)\n",
    "    if y is not None:\n",
    "        y_embd = to_categorical(y, np.unique(y).size)\n",
    "        \n",
    "        return x_mirna_embd, x_mrna_embd, y_embd\n",
    "    else:\n",
    "        return x_mirna_embd, x_mrna_embd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[[0. 0. 0. 1.]\n",
      "  [0. 0. 0. 1.]\n",
      "  [0. 1. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[0. 0. 1. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [0. 0. 1. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[1. 0. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [0. 0. 0. 1.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " ...\n",
      "\n",
      " [[0. 1. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[0. 0. 1. 0.]\n",
      "  [0. 0. 1. 0.]\n",
      "  [0. 0. 1. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[1. 0. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]] [[[1. 0. 0. 0.]\n",
      "  [0. 0. 1. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 1.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [1. 0. 0. 0.]]\n",
      "\n",
      " [[0. 0. 1. 0.]\n",
      "  [0. 1. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[1. 0. 0. 0.]\n",
      "  [0. 0. 0. 1.]\n",
      "  [0. 0. 1. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " ...\n",
      "\n",
      " [[0. 0. 1. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [0. 0. 1. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[0. 0. 1. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  [0. 1. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]\n",
      "\n",
      " [[0. 1. 0. 0.]\n",
      "  [0. 1. 0. 0.]\n",
      "  [1. 0. 0. 0.]\n",
      "  ...\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]\n",
      "  [0. 0. 0. 0.]]]\n"
     ]
    }
   ],
   "source": [
    "x_mirna_embd, x_mrna_embd=preprocess_data(mirna_seqs[:10], mrna_seqs[:10])\n",
    "\n",
    "print (x_mirna_embd, x_mrna_embd)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TJ7F04It3qZn"
   },
   "source": [
    "Let's now define how matching can happen.\n",
    "Following the paper, I utilize relaxed site patterns, which covers most of the canonical site types (CSTs), non-canonical site types (NSTs), and context-dependent non-canonical site types (CDNSTs, to define the candidate target site (CTS). The details used are as follows:\n",
    "* 10-mer-m6: six WC pairings from the miRNA nucleotides 1–10\n",
    "* 10-mer-m7: seven WC pairings from the miRNA nucleotides 1–10\n",
    "* Offset 9-mer-m7: seven WC pairings from the miRNA nucleotides 2–10\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "id": "99eN8pN65qad"
   },
   "outputs": [],
   "source": [
    "def find_candidate(mirna_sequence, mrna_sequence, seed_match):\n",
    "    positions = set()\n",
    "    \n",
    "    if seed_match == '10-mer-m6':\n",
    "        SEED_START = 1\n",
    "        SEED_END = 10\n",
    "        SEED_OFFSET = SEED_START - 1\n",
    "        MIN_MATCH = 6\n",
    "        TOLERANCE = (SEED_END-SEED_START+1) - MIN_MATCH\n",
    "    elif seed_match == '10-mer-m7':\n",
    "        SEED_START = 1\n",
    "        SEED_END = 10\n",
    "        SEED_OFFSET = SEED_START - 1\n",
    "        MIN_MATCH = 7\n",
    "        TOLERANCE = (SEED_END-SEED_START+1) - MIN_MATCH\n",
    "    elif seed_match == 'offset-9-mer-m7':\n",
    "        SEED_START = 2\n",
    "        SEED_END = 10\n",
    "        SEED_OFFSET = SEED_START - 1\n",
    "        MIN_MATCH = 7\n",
    "        TOLERANCE = (SEED_END-SEED_START+1) - MIN_MATCH\n",
    "    elif seed_match == 'strict':\n",
    "        positions = find_strict_candidate(mirna_sequence, mrna_sequence)\n",
    "        \n",
    "        return positions\n",
    "    else:\n",
    "        raise ValueError(\"seed_match expected 'strict', '10-mer-m6', '10-mer-m7', or 'offset-9-mer-m7', got '{}'\".format(seed_match))\n",
    "    \n",
    "    seed = mirna_sequence[(SEED_START-1):SEED_END]\n",
    "    rc_seed = str(Seq(seed).complement())\n",
    "    match_iter = regex.finditer(\"({}){{e<={}}}\".format(rc_seed, TOLERANCE), mrna_sequence)\n",
    "    \n",
    "    for match_index in match_iter:\n",
    "        #positions.add(match_index.start()) # slice-start indicies\n",
    "        positions.add(match_index.end()+SEED_OFFSET) # slice-stop indicies\n",
    "    \n",
    "    positions = list(positions)\n",
    "    \n",
    "    return positions\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "id": "qReji6mOehUD"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[64, 161, 130, 185, 225, 322, 232, 202, 13, 302, 143, 112, 53, 342, 216, 25, 123]\n"
     ]
    }
   ],
   "source": [
    "positions=find_candidate(mirna_seqs[1], mrna_seqs[1], seed_match = '10-mer-m6')\n",
    "print(positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_strict_candidate(mirna_sequence, mrna_sequence):\n",
    "    #find position of matches without tolerance\n",
    "\n",
    "    positions = set()\n",
    "    \n",
    "    SEED_TYPES = ['8-mer', '7-mer-m8', '7-mer-A1', '6-mer', '6-mer-A1', 'offset-7-mer', 'offset-6-mer']\n",
    "    for seed_match in SEED_TYPES:\n",
    "        if seed_match == '8-mer':\n",
    "            SEED_START = 2\n",
    "            SEED_END = 8\n",
    "            SEED_OFFSET = 0\n",
    "            seed = 'U' + mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == '7-mer-m8':\n",
    "            SEED_START = 1\n",
    "            SEED_END = 8\n",
    "            SEED_OFFSET = 0\n",
    "            seed = mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == '7-mer-A1':\n",
    "            SEED_START = 2\n",
    "            SEED_END = 7\n",
    "            SEED_OFFSET = 0\n",
    "            seed = 'U' + mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == '6-mer':\n",
    "            SEED_START = 2\n",
    "            SEED_END = 7\n",
    "            SEED_OFFSET = 1\n",
    "            seed = mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == '6mer-A1':\n",
    "            SEED_START = 2\n",
    "            SEED_END = 6\n",
    "            SEED_OFFSET = 0\n",
    "            seed = 'U' + mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == 'offset-7-mer':\n",
    "            SEED_START = 3\n",
    "            SEED_END = 9\n",
    "            SEED_OFFSET = 0\n",
    "            seed = mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        elif seed_match == 'offset-6-mer':\n",
    "            SEED_START = 3\n",
    "            SEED_END = 8\n",
    "            SEED_OFFSET = 0\n",
    "            seed = mirna_sequence[(SEED_START-1):SEED_END]\n",
    "        \n",
    "        rc_seed = str(Seq(seed).complement())\n",
    "        match_iter = regex.finditer(rc_seed, mrna_sequence)\n",
    "        \n",
    "        for match_index in match_iter:\n",
    "            #positions.add(match_index.start()) # slice-start indicies\n",
    "            positions.add(match_index.end()+SEED_OFFSET) # slice-stop indicies\n",
    "    \n",
    "    positions = list(positions)\n",
    "    \n",
    "    return positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[209]\n"
     ]
    }
   ],
   "source": [
    "positions=find_strict_candidate(mirna_seqs[6], mrna_seqs[1])\n",
    "print(positions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_candidate(mirna_sequence, mrna_sequence, cts_size, seed_match):\n",
    "    #using the find_candidate function we can find actual candidates and positions\n",
    "    positions = find_candidate(mirna_sequence, mrna_sequence, seed_match)\n",
    "    \n",
    "    candidates = []\n",
    "    for i in positions:\n",
    "        site_sequence = mrna_sequence[max(0, i-cts_size):i]\n",
    "        rev_site_sequence = site_sequence[::-1]\n",
    "        rc_site_sequence = str(Seq(rev_site_sequence).complement())\n",
    "        candidates.append(rev_site_sequence) # miRNAs: 5'-ends to 3'-ends,  mRNAs: 3'-ends to 5'-ends\n",
    "        #candidates.append(rc_site_sequence)\n",
    "    \n",
    "    return candidates, positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C 64\n",
      "U 161\n",
      "U 130\n",
      "U 185\n",
      "U 225\n",
      "U 322\n",
      "U 232\n",
      "U 202\n",
      "C 13\n",
      "U 302\n",
      "G 143\n",
      "U 112\n",
      "U 53\n",
      "G 342\n",
      "A 216\n",
      "U 25\n",
      "G 123\n"
     ]
    }
   ],
   "source": [
    "candidates, positions=get_candidate(mirna_seqs[1], mrna_seqs[1], cts_size=1, seed_match = '10-mer-m6')\n",
    "for i in range(len(candidates)):\n",
    "    print(candidates[i], positions [i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "id": "s9MHQcIDa5EB"
   },
   "outputs": [],
   "source": [
    "def make_pair(mirna_sequence, mrna_sequence, cts_size, seed_match):\n",
    "    #and finally identify mirna_querys and mrna_targets\n",
    "    candidates, positions = get_candidate(mirna_sequence, mrna_sequence, cts_size, seed_match)\n",
    "    \n",
    "    mirna_querys = []\n",
    "    mrna_targets = []\n",
    "    if len(candidates) == 0:\n",
    "        return (mirna_querys, mrna_targets, positions)\n",
    "    else:\n",
    "        mirna_sequence = mirna_sequence[0:cts_size]\n",
    "        for i in range(len(candidates)):\n",
    "            mirna_querys.append(mirna_sequence)\n",
    "            mrna_targets.append(candidates[i])\n",
    "        \n",
    "    return mirna_querys, mrna_targets, positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "G C 64\n",
      "G U 161\n",
      "G U 130\n",
      "G U 185\n",
      "G U 225\n",
      "G U 322\n",
      "G U 232\n",
      "G U 202\n",
      "G C 13\n",
      "G U 302\n",
      "G G 143\n",
      "G U 112\n",
      "G U 53\n",
      "G G 342\n",
      "G A 216\n",
      "G U 25\n",
      "G G 123\n"
     ]
    }
   ],
   "source": [
    "mirna_querys, mrna_targets, positions = make_pair(mirna_seqs[1], mrna_seqs[1], cts_size=1, seed_match = '10-mer-m6')\n",
    "for i in range(len(candidates)):\n",
    "    print(mirna_querys[i], mrna_targets[i], positions [i])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UENUtYs30IxG"
   },
   "source": [
    "We also need some functions to read the training and test files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hsa-let-7a-5p ENSG00000114573 ATP6V1A\n",
      "hsa-let-7a-5p ENSG00000104497 SNX16\n",
      "hsa-let-7a-5p ENSG00000141682 PMAIP1\n",
      "hsa-let-7a-5p ENSG00000174010 KLHL15\n",
      "hsa-let-7b-5p ENSG00000130402 ACTN4\n",
      "hsa-let-7c-5p ENSG00000130402 ACTN4\n",
      "hsa-let-7a-5p ENSG00000130402 ACTN4\n",
      "hsa-let-7d ENSG00000130402 ACTN4\n",
      "hsa-let-7b-5p ENSG00000105202 FBL\n",
      "hsa-let-7c-5p ENSG00000105202 FBL\n"
     ]
    }
   ],
   "source": [
    "def read_ground_truth(ground_truth_file, header=True, train=True):\n",
    "    # read the trainign and test files containing pairs of miRNA-mRNA \n",
    "    # input format: [miRNA_ID, mRNA_ID, LABEL]\n",
    "    if header is True:\n",
    "        records = pd.read_csv(ground_truth_file, header=0, sep='\\t')\n",
    "    else:\n",
    "        records = pd.read_csv(ground_truth_file, header=None, sep='\\t')\n",
    "    \n",
    "    query_ids = np.asarray(records.iloc[:, 0].values)\n",
    "    target_ids = np.asarray(records.iloc[:, 1].values)\n",
    "    if train is True:\n",
    "        labels = np.asarray(records.iloc[:, 2].values)\n",
    "    else:\n",
    "        labels = np.full((len(records),), fill_value=-1)\n",
    "    \n",
    "    return query_ids, target_ids, labels\n",
    "\n",
    "ground_truth_file = \"data/train/train_set.csv\"\n",
    "query_ids, target_ids, labels=read_ground_truth(ground_truth_file)\n",
    "for i in range(10):\n",
    "    print(query_ids[i], target_ids[i], labels[i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_input_pair(mirna_fasta_file, mrna_fasta_file, ground_truth_file, cts_size=30, seed_match='offset-9-mer-m7', header=True, train=True):\n",
    "    #from sequences, ids and ground truth we generate the dataset\n",
    "    mirna_ids, mirna_seqs, mrna_ids, mrna_seqs = read_fasta(mirna_fasta_file, mrna_fasta_file)\n",
    "    query_ids, target_ids, labels = read_ground_truth(ground_truth_file, header=header, train=train)\n",
    "    \n",
    "    dataset = {\n",
    "        'mirna_fasta_file': mirna_fasta_file,\n",
    "        'mrna_fasta_file': mrna_fasta_file,\n",
    "        'ground_truth_file': ground_truth_file,\n",
    "        'query_ids': [],\n",
    "        'query_seqs': [],\n",
    "        'target_ids': [],\n",
    "        'target_seqs': [],\n",
    "        'target_locs': [],\n",
    "        'labels': []\n",
    "    }\n",
    "    \n",
    "    for i in range(len(query_ids)):\n",
    "        try:\n",
    "            j = mirna_ids.index(query_ids[i])\n",
    "        except ValueError:\n",
    "            continue\n",
    "        try:\n",
    "            k = mrna_ids.index(target_ids[i])\n",
    "        except ValueError:\n",
    "            continue\n",
    "        \n",
    "        query_seqs, target_seqs, locations = make_pair(mirna_seqs[j], mrna_seqs[k], cts_size=cts_size, seed_match=seed_match)\n",
    "        \n",
    "        n_pairs = len(locations)\n",
    "        if n_pairs > 0:\n",
    "            queries = [query_ids[i] for n in range(n_pairs)]\n",
    "            dataset['query_ids'].extend(queries)\n",
    "            dataset['query_seqs'].extend(query_seqs)\n",
    "            \n",
    "            targets = [target_ids[i] for n in range(n_pairs)]\n",
    "            dataset['target_ids'].extend(targets)\n",
    "            dataset['target_seqs'].extend(target_seqs)\n",
    "            dataset['target_locs'].extend(locations)\n",
    "            \n",
    "            dataset['labels'].extend([[labels[i]] for p in range(n_pairs)])\n",
    "        \n",
    "    return dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset=make_input_pair(mirna_fasta_file, mrna_fasta_file, ground_truth_file, seed_match='offset-9-mer-m7')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "UEnv0A1jxz9Z"
   },
   "source": [
    "Let's now prepare the dataset using the above basic functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "id": "G_wDFGrOx0Zi"
   },
   "outputs": [],
   "source": [
    "def get_negative_pair(mirna_fasta_file, mrna_fasta_file, ground_truth_file=None, cts_size=30, seed_match='offset-9-mer-m7', header=False, predict_mode=True):\n",
    "    mirna_ids, mirna_seqs, mrna_ids, mrna_seqs = read_fasta(mirna_fasta_file, mrna_fasta_file)\n",
    "    \n",
    "    dataset = {\n",
    "        'query_ids': [],\n",
    "        'target_ids': [],\n",
    "        'predicts': []\n",
    "    }\n",
    "    \n",
    "    if ground_truth_file is not None:\n",
    "        query_ids, target_ids, labels = read_ground_truth(ground_truth_file, header=header)\n",
    "        \n",
    "        for i in range(len(query_ids)):\n",
    "            try:\n",
    "                j = mirna_ids.index(query_ids[i])\n",
    "            except ValueError:\n",
    "                continue\n",
    "            try:\n",
    "                k = mrna_ids.index(target_ids[i])\n",
    "            except ValueError:\n",
    "                continue\n",
    "\n",
    "            query_seqs, target_seqs, locations = make_pair(mirna_seqs[j], mrna_seqs[k], cts_size=cts_size, seed_match=seed_match)\n",
    "\n",
    "            n_pairs = len(locations)\n",
    "            if (n_pairs == 0) and (predict_mode is True):\n",
    "                dataset['query_ids'].append(query_ids[i])\n",
    "                dataset['target_ids'].append(target_ids[i])\n",
    "                dataset['predicts'].append(0)\n",
    "            elif (n_pairs == 0) and (predict_mode is False):\n",
    "                dataset['query_ids'].append(query_ids[i])\n",
    "                dataset['target_ids'].append(target_ids[i])\n",
    "                dataset['predicts'].append(labels[i])\n",
    "    else:\n",
    "        for i in range(len(mirna_ids)):\n",
    "            for j in range(len(mrna_ids)):\n",
    "                query_seqs, target_seqs, locations = make_pair(mirna_seqs[i], mrna_seqs[j], cts_size=cts_size, seed_match=seed_match)\n",
    "\n",
    "                n_pairs = len(locations)\n",
    "                if n_pairs == 0:\n",
    "                    dataset['query_ids'].append(mirna_ids[i])                \n",
    "                    dataset['target_ids'].append(mrna_ids[j])\n",
    "                    dataset['predicts'].append(0)\n",
    "    \n",
    "    dataset['target_locs'] = [-1 for i in range(len(dataset['query_ids']))]\n",
    "    dataset['probabilities'] = [0.0 for i in range(len(dataset['query_ids']))]\n",
    "    \n",
    "    return dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_2=get_negative_pair(mirna_fasta_file, mrna_fasta_file, ground_truth_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'query_ids': [],\n",
       " 'target_ids': [],\n",
       " 'predicts': [],\n",
       " 'target_locs': [],\n",
       " 'probabilities': []}"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dataset_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def postprocess_result(dataset, probabilities, predicts, predict_mode=True, output_file=None, cts_size=30, seed_match='offset-9-mer-m7', level='site'):\n",
    "    neg_pairs = get_negative_pair(dataset['mirna_fasta_file'], dataset['mrna_fasta_file'], dataset['ground_truth_file'], cts_size=cts_size, seed_match=seed_match, predict_mode=predict_mode)\n",
    "    \n",
    "    query_ids = np.append(dataset['query_ids'], neg_pairs['query_ids'])\n",
    "    target_ids = np.append(dataset['target_ids'], neg_pairs['target_ids'])\n",
    "    target_locs = np.append(dataset['target_locs'], neg_pairs['target_locs'])\n",
    "    probabilities = np.append(probabilities, neg_pairs['probabilities'])\n",
    "    predicts = np.append(predicts, neg_pairs['predicts'])\n",
    "    \n",
    "    # output format: [QUERY, TARGET, LOCATION, PROBABILITY]\n",
    "    records = pd.DataFrame(columns=['MIRNA_ID', 'MRNA_ID', 'LOCATION', 'PROBABILITY'])\n",
    "    records['MIRNA_ID'] = query_ids\n",
    "    records['MRNA_ID'] = target_ids\n",
    "    records['LOCATION'] = np.array([\"{},{}\".format(max(1, l-cts_size+1), l) if l != -1 else \"-1,-1\" for l in target_locs])\n",
    "    records['PROBABILITY'] = probabilities\n",
    "    if predict_mode is True:\n",
    "        records['PREDICT'] = predicts\n",
    "    else:\n",
    "        records['LABEL'] = predicts\n",
    "    \n",
    "    #records = records.sort_values(by=['PROBABILITY', 'MIRNA_ID', 'MRNA_ID'], ascending=[False, True, True])\n",
    "    unique_records = records.drop_duplicates(subset=['MIRNA_ID', 'MRNA_ID'], keep='first')\n",
    "    \n",
    "    if level == 'site':\n",
    "        if output_file is not None:\n",
    "            records.to_csv(output_file, index=False, sep='\\t')\n",
    "        \n",
    "        return records\n",
    "    elif level == 'gene':\n",
    "        if output_file is not None:\n",
    "            unique_records.to_csv(output_file, index=False, sep='\\t')\n",
    "        \n",
    "        return unique_records\n",
    "    else:\n",
    "        raise ValueError(\"level expected 'site' or 'gene', got '{}'\".format(mode))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "records=postprocess_result(dataset, probabilities=None, predicts=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIRNA_ID</th>\n",
       "      <th>MRNA_ID</th>\n",
       "      <th>LOCATION</th>\n",
       "      <th>PROBABILITY</th>\n",
       "      <th>PREDICT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   MIRNA_ID  MRNA_ID  LOCATION PROBABILITY PREDICT\n",
       "0       NaN      NaN       NaN        None    None"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "records"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "dyvNiVOZyvET"
   },
   "source": [
    "Time to setup the dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "id": "TQmlodwsywZH"
   },
   "outputs": [],
   "source": [
    "class Dataset(torch.utils.data.Dataset):\n",
    "    def __init__(self, mirna_fasta_file, mrna_fasta_file, ground_truth_file, cts_size=30, seed_match='offset-9-mer-m7', header=True, train=True):\n",
    "        self.dataset = make_input_pair(mirna_fasta_file, mrna_fasta_file, ground_truth_file, cts_size=cts_size, seed_match=seed_match, header=header, train=train)\n",
    "        self.mirna, self.mrna = preprocess_data(self.dataset['query_seqs'], self.dataset['target_seqs'])\n",
    "        self.labels = np.asarray(self.dataset['labels']).reshape(-1,)\n",
    "        \n",
    "        self.mirna = self.mirna.transpose((0, 2, 1))\n",
    "        self.mrna = self.mrna.transpose((0, 2, 1))\n",
    "        \n",
    "    def __getitem__(self, index):\n",
    "        if torch.is_tensor(index):\n",
    "            index = index.tolist()\n",
    "            \n",
    "        mirna = self.mirna[index]\n",
    "        mrna = self.mrna[index]\n",
    "        label = self.labels[index]\n",
    "        \n",
    "        return (mirna, mrna), label\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.labels)\n",
    "\n",
    "\n",
    "class TrainDataset(torch.utils.data.Dataset):\n",
    "    def __init__(self, ground_truth_file, cts_size=30):\n",
    "        self.records = pd.read_csv(ground_truth_file, header=0, sep='\\t')\n",
    "        mirna_seqs = self.records['MIRNA_SEQ'].values.tolist()\n",
    "        mrna_seqs = self.records['MRNA_SEQ'].values.tolist()\n",
    "        self.mirna, self.mrna = preprocess_data(mirna_seqs, mrna_seqs, cts_size=cts_size)\n",
    "        self.labels = self.records['LABEL'].values.astype(int)\n",
    "        \n",
    "        self.mirna = self.mirna.transpose((0, 2, 1))\n",
    "        self.mrna = self.mrna.transpose((0, 2, 1))\n",
    "    \n",
    "    def __getitem__(self, index):\n",
    "        if torch.is_tensor(index):\n",
    "            index = index.tolist()\n",
    "        \n",
    "        mirna = self.mirna[index]\n",
    "        mrna = self.mrna[index]\n",
    "        label = self.labels[index]\n",
    "        \n",
    "        return (mirna, mrna), label\n",
    "        \n",
    "    def __len__(self):\n",
    "        return len(self.labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "0ESX-Tln01VH"
   },
   "source": [
    "# Model design\n",
    "I'll now build the neural network model and define how it will be trained\n",
    "\n",
    "The proposed approach exploits one-dimensional convolutional neural networks (CNNs) based on sequence-to-sequence interaction learning framework.\n",
    "\n",
    "The one-hot encoded inputs xmi and xm share the same L1 layer, then the concatenated sequence representation h from the L1 layer are fed to the consecutive L2 –L4 layers. The feature maps from the L4 layer will be flattened and fed to L5 layer, then the L6 layer gets logits of the given miRNA-mRNA pair.\n",
    "\n",
    "For training, I will optimize the weighted cross-entropy loss function using Adam optimizer.\n",
    "\n",
    "<img src='deepTarget_model_layout.png' style='width:800px' align='left' />\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "id": "uXgFZyFd0nZJ"
   },
   "outputs": [],
   "source": [
    "class HyperParam:\n",
    "    def __init__(self, filters=None, kernels=None, model_json=None):\n",
    "        self.dictionary = dict()\n",
    "        self.name_postfix = str()\n",
    "\n",
    "        if (filters is not None) and (kernels is not None) and (model_json is None):\n",
    "            for i, (f, k) in enumerate(zip(filters, kernels)):\n",
    "                setattr(self, 'f{}'.format(i+1), f)\n",
    "                setattr(self, 'k{}'.format(i+1), k)\n",
    "                self.dictionary.update({'f{}'.format(i+1): f, 'k{}'.format(i+1): k})\n",
    "            self.len = i+1\n",
    "                \n",
    "            for key, value in self.dictionary.items():\n",
    "                self.name_postfix = \"{}_{}-{}\".format(self.name_postfix, key, value)\n",
    "        elif model_json is not None:\n",
    "            self.dictionary = json.loads(model_json)\n",
    "            for i, (key, value) in enumerate(self.dictionary.items()):\n",
    "                setattr(self, key, value)\n",
    "                self.name_postfix = \"{}_{}-{}\".format(self.name_postfix, key, value)\n",
    "            self.len = (i+1)//2\n",
    "    \n",
    "    def __len__(self):\n",
    "        return self.len\n",
    "\n",
    "\n",
    "class deepTarget(nn.Module):\n",
    "    #defining the architecture of the convonutional model\n",
    "    def __init__(self, hparams=None, hidden_units=30, input_shape=(1, 4, 30), name_prefix=\"model\"):\n",
    "        super(deepTarget, self).__init__()\n",
    "        \n",
    "        if hparams is None:\n",
    "            filters, kernels = [32, 16, 64, 16], [3, 3, 3, 3]\n",
    "            hparams = HyperParam(filters, kernels)\n",
    "        self.name = \"{}{}\".format(name_prefix, hparams.name_postfix)\n",
    "        \n",
    "        if (isinstance(hparams, HyperParam)) and (len(hparams) == 4):\n",
    "            self.embd1 = nn.Conv1d(4, hparams.f1, kernel_size=hparams.k1, padding=((hparams.k1 - 1) // 2))\n",
    "            self.conv2 = nn.Conv1d(hparams.f1*2, hparams.f2, kernel_size=hparams.k2)\n",
    "            self.conv3 = nn.Conv1d(hparams.f2, hparams.f3, kernel_size=hparams.k3)\n",
    "            self.conv4 = nn.Conv1d(hparams.f3, hparams.f4, kernel_size=hparams.k4)\n",
    "            \n",
    "            \"\"\" out_features = ((in_length - kernel_size + (2 * padding)) / stride + 1) * out_channels \"\"\"\n",
    "            flat_features = self.forward(torch.rand(input_shape), torch.rand(input_shape), flat_check=True)\n",
    "            self.fc1 = nn.Linear(flat_features, hidden_units)\n",
    "            self.fc2 = nn.Linear(hidden_units, 2)\n",
    "        else:\n",
    "            raise ValueError(\"not enough hyperparameters\")\n",
    "    \n",
    "    def forward(self, x_mirna, x_mrna, flat_check=False):\n",
    "        h_mirna = F.relu(self.embd1(x_mirna))\n",
    "        h_mrna = F.relu(self.embd1(x_mrna))\n",
    "        \n",
    "        h = torch.cat((h_mirna, h_mrna), dim=1)\n",
    "        h = F.relu(self.conv2(h))\n",
    "        h = F.relu(self.conv3(h))\n",
    "        h = F.relu(self.conv4(h))\n",
    "        \n",
    "        h = h.view(h.size(0), -1)\n",
    "        if flat_check:\n",
    "            return h.size(1)\n",
    "        h = self.fc1(h)\n",
    "        y = self.fc2(h) #y = F.softmax(self.fc2(h), dim=1)\n",
    "        \n",
    "        return y\n",
    "    \n",
    "    def size(self):\n",
    "        return sum(p.numel() for p in self.parameters() if p.requires_grad)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "id": "S82-L9LuyyTT"
   },
   "outputs": [],
   "source": [
    "def train_model(mirna_fasta_file, mrna_fasta_file, train_file, model=None, cts_size=30, seed_match='offset-9-mer-m7', level='gene', batch_size=32, epochs=10, save_file=None, device='cpu'):\n",
    "    if not isinstance(model, deepTarget):\n",
    "        raise ValueError(\"'model' expected <nn.Module 'deepTarget'>, got {}\".format(type(model)))\n",
    "    \n",
    "    print(\"\\n[TRAIN] {}\".format(model.name))\n",
    "    \n",
    "    if train_file.split('/')[-1] == 'train_set.csv':\n",
    "        train_set = TrainDataset(train_file)\n",
    "    else:\n",
    "        train_set = Dataset(mirna_fasta_file, mrna_fasta_file, train_file, seed_match=seed_match, header=True, train=True)\n",
    "    train_loader = DataLoader(dataset=train_set, batch_size=batch_size, shuffle=True)\n",
    "    \n",
    "    class_weight = torch.Tensor(compute_class_weight('balanced', classes=np.unique(train_set.labels), y=train_set.labels)).to(device)\n",
    "    criterion = nn.CrossEntropyLoss(weight=class_weight)\n",
    "    optimizer = optim.Adam(model.parameters())\n",
    "    \n",
    "    model = model.to(device)\n",
    "    for epoch in range(epochs):\n",
    "        epoch_loss, corrects = 0, 0\n",
    "\n",
    "        with tqdm(train_loader, desc=\"Epoch {}/{}\".format(epoch+1, epochs), bar_format=bar_format) as tqdm_loader:\n",
    "            for i, ((mirna, mrna), label) in enumerate(tqdm_loader):\n",
    "                mirna, mrna, label = mirna.to(device, dtype=torch.float), mrna.to(device, dtype=torch.float), label.to(device)\n",
    "                \n",
    "                outputs = model(mirna, mrna)\n",
    "                loss = criterion(outputs, label)\n",
    "                \n",
    "                optimizer.zero_grad()\n",
    "                loss.backward()\n",
    "                optimizer.step()\n",
    "                \n",
    "                epoch_loss += loss.item() * outputs.size(0)\n",
    "                corrects += (torch.max(outputs, 1)[1] == label).sum().item()\n",
    "                \n",
    "                if (i+1) == len(train_loader):\n",
    "                    tqdm_loader.set_postfix(dict(loss=(epoch_loss/len(train_set)), acc=(corrects/len(train_set))))\n",
    "                else:\n",
    "                    tqdm_loader.set_postfix(loss=loss.item())\n",
    "    \n",
    "    if save_file is None:\n",
    "        time = datetime.now()\n",
    "        save_file = \"{}.pt\".format(time.strftime('%Y%m%d_%H%M%S_weights'))\n",
    "    torch.save(model.state_dict(), save_file)\n",
    "    \n",
    "    \n",
    "def predict_result(mirna_fasta_file, mrna_fasta_file, query_file, model=None, weight_file=None, seed_match='offset-9-mer-m7', level='gene', output_file=None, device='cpu'):\n",
    "    if not isinstance(model, deepTarget):\n",
    "        raise ValueError(\"'model' expected <nn.Module 'deepTarget'>, got {}\".format(type(model)))\n",
    "    \n",
    "    if not weight_file.endswith('.pt'):\n",
    "        raise ValueError(\"'weight_file' expected '*.pt', got {}\".format(weight_file))\n",
    "    \n",
    "    model.load_state_dict(torch.load(weight_file))\n",
    "    \n",
    "    test_set = Dataset(mirna_fasta_file, mrna_fasta_file, query_file, seed_match=seed_match, header=True, train=False)\n",
    "    \n",
    "    model = model.to(device)\n",
    "    with torch.no_grad():\n",
    "        model.eval()\n",
    "        \n",
    "        mirna = torch.from_numpy(test_set.mirna).to(device, dtype=torch.float)\n",
    "        mrna = torch.from_numpy(test_set.mrna).to(device, dtype=torch.float)\n",
    "        label = torch.from_numpy(test_set.labels).to(device)\n",
    "        \n",
    "        outputs = model(mirna, mrna)\n",
    "        _, predicts = torch.max(outputs.data, 1)\n",
    "        probabilities = F.softmax(outputs, dim=1)\n",
    "        \n",
    "        y_probs = probabilities.cpu().numpy()[:, 1]\n",
    "        y_predicts = predicts.cpu().numpy()\n",
    "        y_truth = label.cpu().numpy()\n",
    "        \n",
    "        if output_file is None:\n",
    "            time = datetime.now()\n",
    "            output_file = \"{}.csv\".format(time.strftime('%Y%m%d_%H%M%S_results'))\n",
    "        results = postprocess_result(test_set.dataset, y_probs, y_predicts,\n",
    "                                     seed_match=seed_match, level=level, output_file=output_file)\n",
    "        \n",
    "        print(results)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "XoqQg2fA0j3B"
   },
   "source": [
    "# Training\n",
    "Feeding the model with the dataset to train\n",
    "\n",
    "I have trained the model with multiple batch_sizes and number of epochs - turned out that 3 epochs are enough for this problem and avoid overfitting. However, increasing the batch size to 64 showed a better accuracy. I also tested different seed-matches and 10-mer-m6 resulted to be the best one to predict miRNA-mRNA targets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "id": "uX0sgJZ7lqCO"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[START] 2021-03-09 @ 18:44:47\n",
      "\n",
      "[TRAIN] model_f1-32_k1-3_f2-16_k2-3_f3-64_k3-3_f4-16_k4-3\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "f2d90dc6cc1e4319a78c251154eed8b8",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Epoch 1/3', max=1023.0, style=ProgressStyle(description_w…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "9a1194cda42249cdad7af3937e2ed6bd",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Epoch 2/3', max=1023.0, style=ProgressStyle(description_w…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3da7744e7e504c7e9a9ffdc5bdc5a728",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, description='Epoch 3/3', max=1023.0, style=ProgressStyle(description_w…"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "[FINISH] 2021-03-09 @ 18:45:31 (user time: 0:00:43.989682)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "mirna_fasta_file = \"data/mirna.fasta\"\n",
    "mrna_fasta_file  = \"data/mrna.fasta\"\n",
    "level            = \"gene\"\n",
    "train_file       = \"data/train/train_set.csv\"\n",
    "save_file        = \"weights.pt\"\n",
    "seed_match       = \"10-mer-m6\"\n",
    "batch_size       = 64\n",
    "epochs           = 3        \n",
    "device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')\n",
    "\n",
    "start_time = datetime.now()\n",
    "print(\"\\n[START] {}\".format(start_time.strftime('%Y-%m-%d @ %H:%M:%S')))\n",
    "\n",
    "model = deepTarget()\n",
    "train_model(mirna_fasta_file, mrna_fasta_file, train_file,\n",
    "            model=model,\n",
    "            seed_match=seed_match, level=level,\n",
    "            batch_size=batch_size, epochs=epochs,\n",
    "            save_file=save_file, device=device)\n",
    "\n",
    "finish_time = datetime.now()\n",
    "print(\"\\n[FINISH] {} (user time: {})\\n\".format(finish_time.now().strftime('%Y-%m-%d @ %H:%M:%S'), (finish_time - start_time)))\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Predict\n",
    "Using the test set provided, I feed the model with new data to verify predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "[START] 2021-03-09 @ 18:45:31\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "60205   hsa-miR-214-3p     NM_018254      -1,-1     0.000000        0\n",
      "60206     hsa-miR-4523  NM_001164391      -1,-1     0.000000        0\n",
      "60207  hsa-miR-4690-5p     NM_012265      -1,-1     0.000000        0\n",
      "60208   hsa-miR-497-5p     NM_004662      -1,-1     0.000000        0\n",
      "60209      hsa-miR-922     NM_006923      -1,-1     0.000000        0\n",
      "\n",
      "[803 rows x 5 columns]\n",
      "\n",
      "[FINISH] 2021-03-09 @ 18:46:01 (user time: 0:00:30.092920)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "query_file       = \"data/test/test_split_2.csv\"\n",
    "weight_file      = 'weights.pt'\n",
    "output_file      = \"results.csv\"\n",
    "\n",
    "start_time = datetime.now()\n",
    "print(\"\\n[START] {}\".format(start_time.strftime('%Y-%m-%d @ %H:%M:%S')))\n",
    "\n",
    "model = deepTarget()\n",
    "results = predict_result(mirna_fasta_file, mrna_fasta_file, query_file,\n",
    "                         model=model, weight_file=weight_file,\n",
    "                         seed_match=seed_match, level=level,\n",
    "                         output_file=output_file, device=device)\n",
    "\n",
    "finish_time = datetime.now()\n",
    "print(\"\\n[FINISH] {} (user time: {})\\n\".format(finish_time.now().strftime('%Y-%m-%d @ %H:%M:%S'), (finish_time - start_time)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Evaluate\n",
    "Since the dataset is balanced, the performance of the model against the benchmark will be evaluated using common binary classification performance metrics including accuracy [(TP + TN)/(TP + TN + FP + FN), where TP, FP, FN, and TN represent the numbers of true positives, false positives, false negatives, and true negatives, respectively], sensitivity [TP/(TP + FN)], specificity [TN/(TN + FP)], positive predictive value (PPV)[TP/(TP + FP)] and negative predictive value (NPV) [TN/(TN + FN)]. I will finally benchmark against the paper."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Creating pandas dataframes for the predictions and the actual test set\n",
    "predictions=pd.read_csv(\"results.csv\", header=0, sep='\\t')\n",
    "test_set=pd.read_csv(\"data/test/test_split_2.csv\", header=None, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "#addign a key 'miRNA-mRNA' to merge the df\n",
    "key=[]\n",
    "for i in range (len (predictions['MIRNA_ID'])):\n",
    "    key.append(predictions['MIRNA_ID'][i]+predictions['MRNA_ID'][i])\n",
    "predictions['key']=key\n",
    "\n",
    "key2=[]\n",
    "for i in range (len (test_set[0])):\n",
    "    key2.append(test_set[0][i]+test_set[1][i])\n",
    "test_set['key']=key2\n",
    "test_set.index=key2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIRNA_ID</th>\n",
       "      <th>MRNA_ID</th>\n",
       "      <th>LOCATION</th>\n",
       "      <th>PROBABILITY</th>\n",
       "      <th>PREDICT</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>1511,1540</td>\n",
       "      <td>0.960640</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7c-5pNM_005373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-miR-106b-3p</td>\n",
       "      <td>NM_001099285</td>\n",
       "      <td>484,513</td>\n",
       "      <td>0.362340</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-106b-3pNM_001099285</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-miR-107</td>\n",
       "      <td>NM_002970</td>\n",
       "      <td>131,160</td>\n",
       "      <td>0.997281</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-107NM_002970</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-miR-10a-5p</td>\n",
       "      <td>NM_001105541</td>\n",
       "      <td>227,256</td>\n",
       "      <td>0.192497</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-10a-5pNM_001105541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-miR-10a-5p</td>\n",
       "      <td>NM_001166687</td>\n",
       "      <td>356,385</td>\n",
       "      <td>0.551413</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-10a-5pNM_001166687</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>hsa-miR-214-3p</td>\n",
       "      <td>NM_018254</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-214-3pNM_018254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>hsa-miR-4523</td>\n",
       "      <td>NM_001164391</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-4523NM_001164391</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>hsa-miR-4690-5p</td>\n",
       "      <td>NM_012265</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-4690-5pNM_012265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>hsa-miR-497-5p</td>\n",
       "      <td>NM_004662</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-497-5pNM_004662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>hsa-miR-922</td>\n",
       "      <td>NM_006923</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-922NM_006923</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>803 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT  \\\n",
       "0      hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1   \n",
       "1    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0   \n",
       "2        hsa-miR-107     NM_002970    131,160     0.997281        1   \n",
       "3     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0   \n",
       "4     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1   \n",
       "..               ...           ...        ...          ...      ...   \n",
       "798   hsa-miR-214-3p     NM_018254      -1,-1     0.000000        0   \n",
       "799     hsa-miR-4523  NM_001164391      -1,-1     0.000000        0   \n",
       "800  hsa-miR-4690-5p     NM_012265      -1,-1     0.000000        0   \n",
       "801   hsa-miR-497-5p     NM_004662      -1,-1     0.000000        0   \n",
       "802      hsa-miR-922     NM_006923      -1,-1     0.000000        0   \n",
       "\n",
       "                             key  \n",
       "0         hsa-let-7c-5pNM_005373  \n",
       "1    hsa-miR-106b-3pNM_001099285  \n",
       "2           hsa-miR-107NM_002970  \n",
       "3     hsa-miR-10a-5pNM_001105541  \n",
       "4     hsa-miR-10a-5pNM_001166687  \n",
       "..                           ...  \n",
       "798      hsa-miR-214-3pNM_018254  \n",
       "799     hsa-miR-4523NM_001164391  \n",
       "800     hsa-miR-4690-5pNM_012265  \n",
       "801      hsa-miR-497-5pNM_004662  \n",
       "802         hsa-miR-922NM_006923  \n",
       "\n",
       "[803 rows x 6 columns]"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "predictions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>key</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>hsa-let-7a-5pNM_005373</th>\n",
       "      <td>hsa-let-7a-5p</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7a-5pNM_005373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7b-5pNM_001291429</th>\n",
       "      <td>hsa-let-7b-5p</td>\n",
       "      <td>NM_001291429</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7b-5pNM_001291429</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-let-7c-5pNM_005373</th>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-let-7c-5pNM_005373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-1-3pNM_001292001</th>\n",
       "      <td>hsa-miR-1-3p</td>\n",
       "      <td>NM_001292001</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1-3pNM_001292001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-1-3pNM_001354579</th>\n",
       "      <td>hsa-miR-1-3p</td>\n",
       "      <td>NM_001354579</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-1-3pNM_001354579</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-98-5pNM_001144999</th>\n",
       "      <td>hsa-miR-98-5p</td>\n",
       "      <td>NM_001144999</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-98-5pNM_001144999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-98-5pNM_001256310</th>\n",
       "      <td>hsa-miR-98-5p</td>\n",
       "      <td>NM_001256310</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-98-5pNM_001256310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-98-5pNM_001284388</th>\n",
       "      <td>hsa-miR-98-5p</td>\n",
       "      <td>NM_001284388</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-98-5pNM_001284388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-98-5pNM_006662</th>\n",
       "      <td>hsa-miR-98-5p</td>\n",
       "      <td>NM_006662</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-98-5pNM_006662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>hsa-miR-98-5pNM_033081</th>\n",
       "      <td>hsa-miR-98-5p</td>\n",
       "      <td>NM_033081</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-98-5pNM_033081</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1095 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                       0             1  2  \\\n",
       "hsa-let-7a-5pNM_005373     hsa-let-7a-5p     NM_005373  0   \n",
       "hsa-let-7b-5pNM_001291429  hsa-let-7b-5p  NM_001291429  0   \n",
       "hsa-let-7c-5pNM_005373     hsa-let-7c-5p     NM_005373  0   \n",
       "hsa-miR-1-3pNM_001292001    hsa-miR-1-3p  NM_001292001  0   \n",
       "hsa-miR-1-3pNM_001354579    hsa-miR-1-3p  NM_001354579  0   \n",
       "...                                  ...           ... ..   \n",
       "hsa-miR-98-5pNM_001144999  hsa-miR-98-5p  NM_001144999  1   \n",
       "hsa-miR-98-5pNM_001256310  hsa-miR-98-5p  NM_001256310  1   \n",
       "hsa-miR-98-5pNM_001284388  hsa-miR-98-5p  NM_001284388  1   \n",
       "hsa-miR-98-5pNM_006662     hsa-miR-98-5p     NM_006662  1   \n",
       "hsa-miR-98-5pNM_033081     hsa-miR-98-5p     NM_033081  1   \n",
       "\n",
       "                                                 key  \n",
       "hsa-let-7a-5pNM_005373        hsa-let-7a-5pNM_005373  \n",
       "hsa-let-7b-5pNM_001291429  hsa-let-7b-5pNM_001291429  \n",
       "hsa-let-7c-5pNM_005373        hsa-let-7c-5pNM_005373  \n",
       "hsa-miR-1-3pNM_001292001    hsa-miR-1-3pNM_001292001  \n",
       "hsa-miR-1-3pNM_001354579    hsa-miR-1-3pNM_001354579  \n",
       "...                                              ...  \n",
       "hsa-miR-98-5pNM_001144999  hsa-miR-98-5pNM_001144999  \n",
       "hsa-miR-98-5pNM_001256310  hsa-miR-98-5pNM_001256310  \n",
       "hsa-miR-98-5pNM_001284388  hsa-miR-98-5pNM_001284388  \n",
       "hsa-miR-98-5pNM_006662        hsa-miR-98-5pNM_006662  \n",
       "hsa-miR-98-5pNM_033081        hsa-miR-98-5pNM_033081  \n",
       "\n",
       "[1095 rows x 4 columns]"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test_set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "#adding the actual label to the predictions dataset\n",
    "labels=[]\n",
    "for i in predictions['key']:\n",
    "    labels.append(test_set.loc[i, 2])\n",
    "predictions['labels']=labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "#givign a label to correct and incorrect results\n",
    "result=[]\n",
    "for i in range(len(predictions['key'])):\n",
    "    if predictions['PREDICT'][i]==predictions['labels'][i]==1:\n",
    "        result.append(\"TP\")\n",
    "    elif predictions['PREDICT'][i]==predictions['labels'][i]==0:\n",
    "        result.append(\"TN\")\n",
    "    elif predictions['labels'][i]==1:\n",
    "        result.append(\"FN\")\n",
    "    else:\n",
    "        result.append(\"FP\")\n",
    "predictions['result']=result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>MIRNA_ID</th>\n",
       "      <th>MRNA_ID</th>\n",
       "      <th>LOCATION</th>\n",
       "      <th>PROBABILITY</th>\n",
       "      <th>PREDICT</th>\n",
       "      <th>key</th>\n",
       "      <th>labels</th>\n",
       "      <th>result</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>hsa-let-7c-5p</td>\n",
       "      <td>NM_005373</td>\n",
       "      <td>1511,1540</td>\n",
       "      <td>0.960640</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-let-7c-5pNM_005373</td>\n",
       "      <td>0</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>hsa-miR-106b-3p</td>\n",
       "      <td>NM_001099285</td>\n",
       "      <td>484,513</td>\n",
       "      <td>0.362340</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-106b-3pNM_001099285</td>\n",
       "      <td>0</td>\n",
       "      <td>TN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>hsa-miR-107</td>\n",
       "      <td>NM_002970</td>\n",
       "      <td>131,160</td>\n",
       "      <td>0.997281</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-107NM_002970</td>\n",
       "      <td>0</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>hsa-miR-10a-5p</td>\n",
       "      <td>NM_001105541</td>\n",
       "      <td>227,256</td>\n",
       "      <td>0.192497</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-10a-5pNM_001105541</td>\n",
       "      <td>0</td>\n",
       "      <td>TN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hsa-miR-10a-5p</td>\n",
       "      <td>NM_001166687</td>\n",
       "      <td>356,385</td>\n",
       "      <td>0.551413</td>\n",
       "      <td>1</td>\n",
       "      <td>hsa-miR-10a-5pNM_001166687</td>\n",
       "      <td>0</td>\n",
       "      <td>FP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>798</th>\n",
       "      <td>hsa-miR-214-3p</td>\n",
       "      <td>NM_018254</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-214-3pNM_018254</td>\n",
       "      <td>1</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>799</th>\n",
       "      <td>hsa-miR-4523</td>\n",
       "      <td>NM_001164391</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-4523NM_001164391</td>\n",
       "      <td>1</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>800</th>\n",
       "      <td>hsa-miR-4690-5p</td>\n",
       "      <td>NM_012265</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-4690-5pNM_012265</td>\n",
       "      <td>1</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>801</th>\n",
       "      <td>hsa-miR-497-5p</td>\n",
       "      <td>NM_004662</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-497-5pNM_004662</td>\n",
       "      <td>1</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>hsa-miR-922</td>\n",
       "      <td>NM_006923</td>\n",
       "      <td>-1,-1</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>hsa-miR-922NM_006923</td>\n",
       "      <td>1</td>\n",
       "      <td>FN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>803 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT  \\\n",
       "0      hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1   \n",
       "1    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0   \n",
       "2        hsa-miR-107     NM_002970    131,160     0.997281        1   \n",
       "3     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0   \n",
       "4     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1   \n",
       "..               ...           ...        ...          ...      ...   \n",
       "798   hsa-miR-214-3p     NM_018254      -1,-1     0.000000        0   \n",
       "799     hsa-miR-4523  NM_001164391      -1,-1     0.000000        0   \n",
       "800  hsa-miR-4690-5p     NM_012265      -1,-1     0.000000        0   \n",
       "801   hsa-miR-497-5p     NM_004662      -1,-1     0.000000        0   \n",
       "802      hsa-miR-922     NM_006923      -1,-1     0.000000        0   \n",
       "\n",
       "                             key  labels result  \n",
       "0         hsa-let-7c-5pNM_005373       0     FP  \n",
       "1    hsa-miR-106b-3pNM_001099285       0     TN  \n",
       "2           hsa-miR-107NM_002970       0     FP  \n",
       "3     hsa-miR-10a-5pNM_001105541       0     TN  \n",
       "4     hsa-miR-10a-5pNM_001166687       0     FP  \n",
       "..                           ...     ...    ...  \n",
       "798      hsa-miR-214-3pNM_018254       1     FN  \n",
       "799     hsa-miR-4523NM_001164391       1     FN  \n",
       "800     hsa-miR-4690-5pNM_012265       1     FN  \n",
       "801      hsa-miR-497-5pNM_004662       1     FN  \n",
       "802         hsa-miR-922NM_006923       1     FN  \n",
       "\n",
       "[803 rows x 8 columns]"
      ]
     },
     "execution_count": 132,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#a look at the final dataset\n",
    "predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To evaluate the metrics, let's see how many true positives, true negatives, false positives and false negatives we got"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "true positives=307, true negatives=329, false positives=88, false negatives=79\n"
     ]
    }
   ],
   "source": [
    "TP=predictions['result'].value_counts()['TP']\n",
    "TN=predictions['result'].value_counts()['TN']\n",
    "FP=predictions['result'].value_counts()['FP']\n",
    "FN=predictions['result'].value_counts()['FN']\n",
    "print('true positives={}, true negatives={}, false positives={}, false negatives={}'.format(TP, TN, FP, FN))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can now calculate the key metrics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "accuracy =  0.7920298879202988\n",
      "sensitivity = 0.7953367875647669\n",
      "specificity = 0.7889688249400479\n",
      "F-measure = 0.7861715749039693\n",
      "PPV = 0.7772151898734178\n",
      "NPV = 0.8063725490196079\n"
     ]
    }
   ],
   "source": [
    "print('accuracy = ',(TP + TN)/(TP + TN + FP + FN))\n",
    "print('sensitivity =', TP/(TP + FN))\n",
    "print('specificity =', TN/(TN + FP))\n",
    "print('F-measure =', 2*TP/(2*TP + FP + FN))\n",
    "print('PPV =', TP/(TP + FP))\n",
    "print('NPV =', TN/(TN + FN))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's now run the same process to all the 10 test sets to generate the final score matrix and bechmark with the paper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_queries=[\"data/test/test_split_0.csv\", \n",
    "              \"data/test/test_split_1.csv\", \n",
    "              \"data/test/test_split_2.csv\", \n",
    "              \"data/test/test_split_3.csv\", \n",
    "              \"data/test/test_split_4.csv\", \n",
    "              \"data/test/test_split_5.csv\", \n",
    "              \"data/test/test_split_6.csv\", \n",
    "              \"data/test/test_split_7.csv\", \n",
    "              \"data/test/test_split_8.csv\", \n",
    "              \"data/test/test_split_9.csv\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "59645   hsa-miR-497-5p  NM_001242825      -1,-1     0.000000        0\n",
      "59646  hsa-miR-548d-3p  NM_001144755      -1,-1     0.000000        0\n",
      "59647  hsa-miR-548d-3p     NM_021127      -1,-1     0.000000        0\n",
      "59648  hsa-miR-548d-3p     NM_024524      -1,-1     0.000000        0\n",
      "59649      hsa-miR-922  NM_001244604      -1,-1     0.000000        0\n",
      "\n",
      "[796 rows x 5 columns]\n",
      "1\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "64112   hsa-miR-505-5p     NM_021141      -1,-1     0.000000        0\n",
      "64113  hsa-miR-526b-5p     NM_022551      -1,-1     0.000000        0\n",
      "64114  hsa-miR-548d-3p     NM_152699      -1,-1     0.000000        0\n",
      "64115  hsa-miR-548h-3p  NM_001161658      -1,-1     0.000000        0\n",
      "64116     hsa-miR-548z     NM_199054      -1,-1     0.000000        0\n",
      "\n",
      "[817 rows x 5 columns]\n",
      "2\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "60205   hsa-miR-214-3p     NM_018254      -1,-1     0.000000        0\n",
      "60206     hsa-miR-4523  NM_001164391      -1,-1     0.000000        0\n",
      "60207  hsa-miR-4690-5p     NM_012265      -1,-1     0.000000        0\n",
      "60208   hsa-miR-497-5p     NM_004662      -1,-1     0.000000        0\n",
      "60209      hsa-miR-922     NM_006923      -1,-1     0.000000        0\n",
      "\n",
      "[803 rows x 5 columns]\n",
      "3\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "60145  hsa-miR-3173-3p     NM_139135      -1,-1     0.000000        0\n",
      "60146   hsa-miR-330-3p  NM_001143826      -1,-1     0.000000        0\n",
      "60147  hsa-miR-3613-3p     NM_032796      -1,-1     0.000000        0\n",
      "60148   hsa-miR-497-5p     NM_002035      -1,-1     0.000000        0\n",
      "60149  hsa-miR-548d-3p  NM_001032284      -1,-1     0.000000        0\n",
      "\n",
      "[813 rows x 5 columns]\n",
      "4\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "60244   hsa-miR-497-5p     NM_032287      -1,-1     0.000000        0\n",
      "60245  hsa-miR-548h-3p     NM_144599      -1,-1     0.000000        0\n",
      "60246  hsa-miR-548h-3p     NM_152699      -1,-1     0.000000        0\n",
      "60247      hsa-miR-922  NM_001135095      -1,-1     0.000000        0\n",
      "60248      hsa-miR-940  NM_001253775      -1,-1     0.000000        0\n",
      "\n",
      "[810 rows x 5 columns]\n",
      "5\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "62695   hsa-miR-330-3p     NM_003478      -1,-1     0.000000        0\n",
      "62696   hsa-miR-497-5p  NM_001199839      -1,-1     0.000000        0\n",
      "62697   hsa-miR-497-5p  NM_001206654      -1,-1     0.000000        0\n",
      "62698  hsa-miR-548h-3p  NM_001172478      -1,-1     0.000000        0\n",
      "62699      hsa-miR-940  NM_001256799      -1,-1     0.000000        0\n",
      "\n",
      "[802 rows x 5 columns]\n",
      "6\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "61033   hsa-miR-506-3p     NM_138325      -1,-1     0.000000        0\n",
      "61034      hsa-miR-941  NM_001145794      -1,-1     0.000000        0\n",
      "61035      hsa-miR-941     NM_015004      -1,-1     0.000000        0\n",
      "61036   hsa-miR-497-3p  NM_001135699      -1,-1     0.000000        0\n",
      "61037     hsa-miR-548z  NM_001130849      -1,-1     0.000000        0\n",
      "\n",
      "[808 rows x 5 columns]\n",
      "7\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "59982  hsa-miR-3613-3p  NM_001080392      -1,-1     0.000000        0\n",
      "59983  hsa-miR-3613-3p  NM_001190274      -1,-1     0.000000        0\n",
      "59984   hsa-miR-497-5p     NM_001539      -1,-1     0.000000        0\n",
      "59985   hsa-miR-497-5p     NM_003129      -1,-1     0.000000        0\n",
      "59986  hsa-miR-548d-3p     NM_015375      -1,-1     0.000000        0\n",
      "\n",
      "[798 rows x 5 columns]\n",
      "8\n",
      "              MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0        hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100    hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113        hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126     hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154     hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                ...           ...        ...          ...      ...\n",
      "60118   hsa-miR-330-3p     NM_031844      -1,-1     0.000000        0\n",
      "60119   hsa-miR-522-5p  NM_001242878      -1,-1     0.000000        0\n",
      "60120  hsa-miR-548d-3p  NM_001206557      -1,-1     0.000000        0\n",
      "60121  hsa-miR-548d-3p     NM_003249      -1,-1     0.000000        0\n",
      "60122      hsa-miR-922  NM_001145357      -1,-1     0.000000        0\n",
      "\n",
      "[800 rows x 5 columns]\n",
      "9\n",
      "               MIRNA_ID       MRNA_ID   LOCATION  PROBABILITY  PREDICT\n",
      "0         hsa-let-7c-5p     NM_005373  1511,1540     0.960640        1\n",
      "100     hsa-miR-106b-3p  NM_001099285    484,513     0.362340        0\n",
      "113         hsa-miR-107     NM_002970    131,160     0.997281        1\n",
      "126      hsa-miR-10a-5p  NM_001105541    227,256     0.192497        0\n",
      "154      hsa-miR-10a-5p  NM_001166687    356,385     0.551413        1\n",
      "...                 ...           ...        ...          ...      ...\n",
      "61518  hsa-miR-548ap-3p  NM_001024843      -1,-1     0.000000        0\n",
      "61519   hsa-miR-548d-3p     NM_003337      -1,-1     0.000000        0\n",
      "61520   hsa-miR-548d-3p     NM_016422      -1,-1     0.000000        0\n",
      "61521      hsa-miR-548y     NM_032194      -1,-1     0.000000        0\n",
      "61522      hsa-miR-548z  NM_001100603      -1,-1     0.000000        0\n",
      "\n",
      "[816 rows x 5 columns]\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "matrix=pd.DataFrame()\n",
    "matrix['set']=['accuracy','sensitivity','specificity','F-measure', 'PPV','NPV']\n",
    "x=0\n",
    "for i in test_queries:\n",
    "    x+=1\n",
    "    output_file = \"results.csv\"\n",
    "    query_file = i\n",
    "    model = deepTarget()\n",
    "    results = predict_result(mirna_fasta_file, mrna_fasta_file, query_file,\n",
    "                         model=model, weight_file=weight_file,\n",
    "                         seed_match=seed_match, level=level,\n",
    "                         output_file=output_file, device=device)\n",
    "    predictions=pd.read_csv(\"results.csv\", header=0, sep='\\t')\n",
    "    test_set=pd.read_csv(i, header=None, sep='\\t')\n",
    "    key=[]\n",
    "    for i in range (len (predictions['MIRNA_ID'])):\n",
    "        key.append(predictions['MIRNA_ID'][i]+predictions['MRNA_ID'][i])\n",
    "    predictions['key']=key\n",
    "\n",
    "    key2=[]\n",
    "    for i in range (len (test_set[0])):\n",
    "        key2.append(test_set[0][i]+test_set[1][i])\n",
    "    test_set['key']=key2\n",
    "    test_set.index=key2\n",
    "    \n",
    "    labels=[]\n",
    "    for i in predictions['key']:\n",
    "        labels.append(test_set.loc[i, 2])\n",
    "    predictions['labels']=labels\n",
    "\n",
    "    result=[]\n",
    "    for i in range(len(predictions['key'])):\n",
    "        if predictions['PREDICT'][i]==predictions['labels'][i]==1:\n",
    "            result.append(\"TP\")\n",
    "        elif predictions['PREDICT'][i]==predictions['labels'][i]==0:\n",
    "            result.append(\"TN\")\n",
    "        elif predictions['labels'][i]==1:\n",
    "            result.append(\"FN\")\n",
    "        else:\n",
    "            result.append(\"FP\")\n",
    "    \n",
    "    predictions['result']=result\n",
    "    TP=predictions['result'].value_counts()['TP']\n",
    "    TN=predictions['result'].value_counts()['TN']\n",
    "    FP=predictions['result'].value_counts()['FP']\n",
    "    FN=predictions['result'].value_counts()['FN']\n",
    "    matrix['set{}'.format(x)]=[(TP + TN)/(TP + TN + FP + FN),\n",
    "              TP/(TP + FN),\n",
    "              TN/(TN + FP),\n",
    "              2*TP/(2*TP + FP + FN),\n",
    "              TP/(TP + FP),\n",
    "              TN/(TN + FN)]\n",
    "    print(x)\n",
    "matrix=matrix.transpose()\n",
    "matrix.columns = matrix.iloc[0]\n",
    "score_matrix=matrix.drop('set')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now let's calculate the average accuracy score ofr the 10 test datasets:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>set</th>\n",
       "      <th>accuracy</th>\n",
       "      <th>sensitivity</th>\n",
       "      <th>specificity</th>\n",
       "      <th>F-measure</th>\n",
       "      <th>PPV</th>\n",
       "      <th>NPV</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>set1</th>\n",
       "      <td>0.792714</td>\n",
       "      <td>0.796834</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.785436</td>\n",
       "      <td>0.774359</td>\n",
       "      <td>0.810345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set2</th>\n",
       "      <td>0.80661</td>\n",
       "      <td>0.825</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.806846</td>\n",
       "      <td>0.789474</td>\n",
       "      <td>0.824561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set3</th>\n",
       "      <td>0.79203</td>\n",
       "      <td>0.795337</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.786172</td>\n",
       "      <td>0.777215</td>\n",
       "      <td>0.806373</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set4</th>\n",
       "      <td>0.789668</td>\n",
       "      <td>0.790404</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.785445</td>\n",
       "      <td>0.780549</td>\n",
       "      <td>0.798544</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set5</th>\n",
       "      <td>0.792593</td>\n",
       "      <td>0.796438</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.788413</td>\n",
       "      <td>0.780549</td>\n",
       "      <td>0.804401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set6</th>\n",
       "      <td>0.783042</td>\n",
       "      <td>0.776623</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.774611</td>\n",
       "      <td>0.77261</td>\n",
       "      <td>0.792771</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set7</th>\n",
       "      <td>0.79703</td>\n",
       "      <td>0.805627</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.793451</td>\n",
       "      <td>0.781638</td>\n",
       "      <td>0.812346</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set8</th>\n",
       "      <td>0.789474</td>\n",
       "      <td>0.790026</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.781818</td>\n",
       "      <td>0.773779</td>\n",
       "      <td>0.804401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set9</th>\n",
       "      <td>0.79375</td>\n",
       "      <td>0.798956</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.787645</td>\n",
       "      <td>0.77665</td>\n",
       "      <td>0.810345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>set10</th>\n",
       "      <td>0.801471</td>\n",
       "      <td>0.814536</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.800493</td>\n",
       "      <td>0.786925</td>\n",
       "      <td>0.816377</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>average</th>\n",
       "      <td>0.793838</td>\n",
       "      <td>0.798978</td>\n",
       "      <td>0.788969</td>\n",
       "      <td>0.789033</td>\n",
       "      <td>0.779375</td>\n",
       "      <td>0.808046</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "set      accuracy sensitivity specificity F-measure       PPV       NPV\n",
       "set1     0.792714    0.796834    0.788969  0.785436  0.774359  0.810345\n",
       "set2      0.80661       0.825    0.788969  0.806846  0.789474  0.824561\n",
       "set3      0.79203    0.795337    0.788969  0.786172  0.777215  0.806373\n",
       "set4     0.789668    0.790404    0.788969  0.785445  0.780549  0.798544\n",
       "set5     0.792593    0.796438    0.788969  0.788413  0.780549  0.804401\n",
       "set6     0.783042    0.776623    0.788969  0.774611   0.77261  0.792771\n",
       "set7      0.79703    0.805627    0.788969  0.793451  0.781638  0.812346\n",
       "set8     0.789474    0.790026    0.788969  0.781818  0.773779  0.804401\n",
       "set9      0.79375    0.798956    0.788969  0.787645   0.77665  0.810345\n",
       "set10    0.801471    0.814536    0.788969  0.800493  0.786925  0.816377\n",
       "average  0.793838    0.798978    0.788969  0.789033  0.779375  0.808046"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#calculating averages\n",
    "averages=[]\n",
    "for i in score_matrix.columns:\n",
    "    averages.append(score_matrix[i].mean())\n",
    "averages=pd.DataFrame(averages)\n",
    "averages.columns=['average']\n",
    "averages['set']=['accuracy','sensitivity','specificity','F-measure', 'PPV','NPV']\n",
    "averages=averages.transpose()\n",
    "averages.columns = averages.iloc[1]\n",
    "averages=averages.drop('set')\n",
    "score_matrix.append(averages)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's compare to the model performance as published in the paper:\n",
    "<img src='evaluation_metrics_deepTarget.png' style='width:800px' align='left'/>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To conclude, this model achives a slightly higher accuracy compared to the deepTarget paper, while reducing the computational effort and time required for training: it gets to +79% accuracy in just 3 epochs, compared to the 10 epochs in the paper."
   ]
  }
 ],
 "metadata": {
  "colab": {
   "collapsed_sections": [],
   "name": "deepTarget project.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "conda_amazonei_pytorch_latest_p36",
   "language": "python",
   "name": "conda_amazonei_pytorch_latest_p36"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
